crate-tmp/src/nvtxw.rs

```diff
@@ -0,0 +1,281 @@
+        return Err(result);
+    }
+    Ok(())
+}
+
+pub fn initialize_simple(backend: Option<OsString>) -> Result<InterfaceHandle, ResultCode> {
+    let default_backend = if cfg!(windows) {
+        "NvtxwBackend.dll"
+    } else if cfg!(target_vendor = "apple") {
+        "libNvtxwBackend.dylib"
+    } else {
+        "libNvtxwBackend.so"
+    };
+
+    let mut path_backend = if let Some(os_backend) = backend {
+        PathBuf::from(os_backend)
+    } else {
+        PathBuf::from(default_backend)
+    };
+    if path_backend.is_dir() {
+        path_backend.push(default_backend);
+    }
+    let s_backend = path_backend.to_str().expect("backend to_str() failed");
+    let c_mode_string = CString::new(s_backend).expect("c_mode_string CString::new failed");
+
+    let mut get_interface_func: GetInterface = Default::default();
+    check(unsafe {
+        nvtxw_bindings::nvtxwInitialize(
+            NVTXW3_INIT_MODE_LIBRARY_FILENAME,
+            c_mode_string.as_ptr(),
+            &mut get_interface_func,
+            null_mut(),
+        )
+    })?;
+
+    let get_interface = get_interface_func.ok_or(NVTXW3_RESULT_FAILED)?;
+
+    let mut interface_void: *const c_void = null();
+    check(unsafe { get_interface(NVTXW3_INTERFACE_ID_CORE_V1, &mut interface_void) })?;
+
+    if interface_void.is_null() {
+        return Err(NVTXW3_RESULT_FAILED);
+    }
+
+    let ptr_interface = interface_void as *const InterfaceHandle;
+    let interface;
+    unsafe {
+        interface = *ptr_interface;
+    }
+
+    Ok(interface)
+}
+
+pub fn session_begin_simple(
+    interface: InterfaceHandle,
+    output: OsString,
+    force: bool,
+    merge: Option<OsString>,
+) -> Result<SessionHandle, ResultCode> {
+    let session_begin = interface
+        .SessionBegin
+        .ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
+
+    let extension = "nsys-rep";
+    let mut path_output = PathBuf::from(output);
+
+    // avoid "*.nsys-rep.nsys-rep", since the backend adds the extension for us
+    if path_output.extension() == Some(OsStr::new(extension)) {
+        path_output.set_extension("");
+    }
+
+    // however, we are responsible for when the output already exists
+    let mut path_test = path_output.with_extension(extension);
+    if path_test.is_dir() {
+        return Err(NVTXW3_RESULT_INVALID_ARGUMENT);
+    } else if path_test.exists() {
+        if force {
+            println!("Overwriting {:?}", &path_test);
+        } else {
+            let mut i = 1;
+            let retry_limit = 100;
+            loop {
+                let e = format!("{}.{}", i, extension);
+                path_test = path_output.with_extension(e);
+                if !path_test.exists() {
+                    path_output.set_extension(format!("{}", i));
+                    break;
+                } else if i >= retry_limit {
+                    return Err(NVTXW3_RESULT_FAILED);
+                }
+                i += 1;
+            }
+        }
+    }
```

Comment:
Unfortunately this code is going to be racy. The equivalent code here is written fairly carefully to make exactly one filesystem API call per attempt:

https://github.com/StanfordLegion/prof-viewer/blob/master/src/archive_data.rs#L21-L46

If I understand correctly, NVTXW will overwrite whatever file you create here. So, I think the safe equivalent of this code is to not ever call `.is_dir()` or `.exists()` but to just go and create the file directly. You can do this with the `File::create_new` API:

https://doc.rust-lang.org/std/fs/struct.File.html#method.create_new

This will fail if the file already exists, so it will not touch any existing files. It will also fail if the path is a directory, so I don't think we need any separate checks for that case.

It's always possible for some other process or user to mess with the file system between when we create this file, and when NVTXW attempts to overwrite it. However, if the process is merely creating files without removing them, then I think this avoids the race. (E.g., think of if the user is running multiple copies of the profiler in parallel.)

Reply:
Ah, I understand! It's like how the only way to know a lock is available is by actually holding it. I've updated the code to use `File::create_new`.

Reply:
I missed that `File::create_new` was added in Rust 1.77. If I understand correctly, the corresponding `OpenOptions` has existed since 1.9, so you can still write this in Rust 1.74, it's just more obnoxious to implement:

```rust
// TODO: replace with File::create_new after Rust 1.77
fn file_create_new<P: AsRef<Path>>(path: P) -> io::Result<File> {
  OpenOptions::new().read(true).write(true).create_new(true).open(path.as_ref())
}
```

And then call `file_create_new` instead of `File::create_new`.

Reply:
Cool, this fixed the CI build.